### ISSUE TITLE
Data snooping on hyperparameter optimization

### ISSUE THREAD

**Issue by vgarciasc** [vgarciasc/hundred-hammers#20]
When doing hyperparameter optimization, every point in the dataset is used; however, some of those points later are used to calculate the test performance.

```python
def evaluate(self, X, y, optim_hyper, n_grid_points):
  if optim_hyper:
      new_models = self.tune_models(X, y, n_grid_points)
  else:
      new_models = deepcopy(self.models)
  
  report, trained_models = self._evaluate_models(X, y, new_models)
  ...
```

 I hadn't thought about this until now, but I think it qualifies as data snooping. I think that the correct way would be to tune a model only on the training data, and use it to evaluate the training and test data. This would involve some non-trivial rewriting, but it seems to be the only correct way to proceed, since the current results with `optim_hyper=True` may not be reproducible for true out-of-sample data. What do you think?

**Comment by eugenioLR** [comment]
Yeah, that makes a lot of sense. The hyperparameters optimization should only be done with the training data, I think i know how to fix it without changing much of the already existing code. I will upload it together with the normalization functionality.
